File: frontend/app/utils/endpoint-identity.ts
export function areEndpointIdentitiesEquivalent(
    first: string | undefined,
    second: string | undefined,
): boolean {
    const firstTrimmed = (first ?? "").trim();
    const secondTrimmed = (second ?? "").trim();
    if (firstTrimmed === secondTrimmed) return true;

    const firstIdentity = createEndpointIdentity(firstTrimmed);
    return firstIdentity !== null && firstIdentity === createEndpointIdentity(secondTrimmed);
}

function createEndpointIdentity(value: string | undefined): string | null {
    const trimmed = (value ?? "").trim();
    const schemeDelimiter = trimmed.indexOf("://");
    if (schemeDelimiter <= 0) return null;

    let endpoint: URL;
    try {
        endpoint = new URL(trimmed);
    } catch {
        return null;
    }
    if (!endpoint.hostname) return null;

    const authorityStart = schemeDelimiter + 3;
    const suffixStart = findSuffixStart(trimmed, authorityStart);
    const authority = trimmed.slice(authorityStart, suffixStart);
    if (authority.includes("\\")) return null;
    const userInfoSeparator = authority.lastIndexOf("@");
    const userInfo = userInfoSeparator < 0 ? "" : authority.slice(0, userInfoSeparator);

    const fragmentStart = trimmed.indexOf("#", suffixStart);
    let queryStart = trimmed.indexOf("?", suffixStart);
    if (fragmentStart >= 0 && queryStart > fragmentStart) queryStart = -1;
    const pathEnd = minNonNegative(trimmed.length, queryStart, fragmentStart);
    const rawPath = trimmed.slice(suffixStart, pathEnd);
    const queryEnd = fragmentStart >= 0 ? fragmentStart : trimmed.length;
    const query = queryStart >= 0 ? trimmed.slice(queryStart, queryEnd) : "";
    const fragment = fragmentStart >= 0 ? trimmed.slice(fragmentStart) : "";

    return JSON.stringify([
        endpoint.protocol.toLowerCase(),
        endpoint.hostname.toLowerCase(),
        endpoint.port || defaultPort(endpoint.protocol),
        userInfo,
        rawPath === "/" ? "" : rawPath,
        query,
        fragment,
    ]);
}

function findSuffixStart(value: string, authorityStart: number): number {
    for (let index = authorityStart; index < value.length; index += 1) {
        if (value[index] === "/" || value[index] === "?" || value[index] === "#") return index;
    }
    return value.length;
}

function minNonNegative(fallback: number, ...values: number[]): number {
    return values.reduce(
        (result, value) => value >= 0 && value < result ? value : result,
        fallback,
    );
}

function defaultPort(protocol: string): string {
    if (protocol.toLowerCase() === "http:") return "80";
    if (protocol.toLowerCase() === "https:") return "443";
    return "";
}

File: backend/Config/EndpointIdentity.cs
namespace NzbWebDAV.Config;

internal readonly record struct EndpointIdentity(
    string Scheme,
    string Host,
    int Port,
    string UserInfo,
    string Path,
    string Query,
    string Fragment)
{
    public static bool AreEquivalent(string? first, string? second)
    {
        return TryCreate(first, out var firstIdentity)
               && TryCreate(second, out var secondIdentity)
               && firstIdentity == secondIdentity;
    }

    private static bool TryCreate(string? value, out EndpointIdentity identity)
    {
        identity = default;
        if (string.IsNullOrWhiteSpace(value)) return false;

        var trimmed = value.Trim();
        var schemeDelimiter = trimmed.IndexOf("://", StringComparison.Ordinal);
        if (schemeDelimiter <= 0
            || !Uri.TryCreate(trimmed, UriKind.Absolute, out var uri)
            || string.IsNullOrWhiteSpace(uri.Host))
        {
            return false;
        }

        var authorityStart = schemeDelimiter + 3;
        var suffixStart = FindSuffixStart(trimmed, authorityStart);
        var authority = trimmed[authorityStart..suffixStart];
        if (authority.Contains('\\')) return false;
        var userInfoSeparator = authority.LastIndexOf('@');
        var userInfo = userInfoSeparator < 0 ? "" : authority[..userInfoSeparator];

        var fragmentStart = trimmed.IndexOf('#', suffixStart);
        var queryStart = trimmed.IndexOf('?', suffixStart);
        if (fragmentStart >= 0 && queryStart > fragmentStart) queryStart = -1;
        var pathEnd = MinNonNegative(trimmed.Length, queryStart, fragmentStart);
        var path = trimmed[suffixStart..pathEnd];
        var queryEnd = fragmentStart >= 0 ? fragmentStart : trimmed.Length;
        var query = queryStart >= 0 ? trimmed[queryStart..queryEnd] : "";
        var fragment = fragmentStart >= 0 ? trimmed[fragmentStart..] : "";

        identity = new EndpointIdentity(
            uri.Scheme.ToLowerInvariant(),
            uri.IdnHost.ToLowerInvariant(),
            uri.Port,
            userInfo,
            path == "/" ? "" : path,
            query,
            fragment);
        return true;
    }

    private static int FindSuffixStart(string value, int authorityStart)
    {
        for (var index = authorityStart; index < value.Length; index++)
        {
            if (value[index] is '/' or '?' or '#') return index;
        }

        return value.Length;
    }

    private static int MinNonNegative(int fallback, params int[] values)
    {
        var result = fallback;
        foreach (var value in values)
        {
            if (value >= 0 && value < result) result = value;
        }

        return result;
    }
}
